examples/palace-tonegen.py

```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env python3
+# Sample for tone generator
+# Copyright (C) 2020  Ngô Ngọc Đức Huy
+#
+# This file is part of palace.
+#
+# palace is free software: you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License as published
+# by the Free Software Foundation, either version 3 of the License,
+# or (at your option) any later version.
+#
+# palace is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with palace.  If not, see <https://www.gnu.org/licenses/>.
+
+from argparse import Action, ArgumentParser
+from array import array
+from math import pi, sin
+from random import random
+from typing import Iterable, Tuple
+
+from palace import Buffer, Context, BaseDecoder, Device
+
+from numpy import linspace
+from numpy.random import random
+from scipy.signal import sawtooth, square, unit_impulse
+
+WAVEFORMS = {'sine': sin,
+             'square': square,
+             'sawtooth': sawtooth,
+             'triangle': lambda time: return sawtooth(time, 0.5),
+             'impulse': unit_impulse,
+             'white noise': random}
+
+
+class ToneGenerator(BaseDecoder):
+    def __init__(self, waveform: str, duration: float, frequency: int):
+        self.func = lambda frame: WAVEFORMS[waveform](frame/self.frequency,
+                                                      frequency)
```

Comment:
Based on the signature, this should be `WAVEFORMS[waveform](frame/self.frequency/frequency)`.  Note that `sin` might need some tweak to have the period of 1.

Reply:
Each of them needs some tweak I believe.

Reply:
Finish the playback part so we can listen to it.

Reply:
`buf.play` doesn't play it?

Anyway, I've got this and don't know what to do.

![image](https://user-images.githubusercontent.com/19296926/79090539-38274f00-7d74-11ea-808f-371f9a530e93.png)